### PR TITLE
Lazyannotator

### PR DESCRIPTION
**Pull request by mssammon**
Changed Annotator to allow lazy initialization. Key is that all users must call "getView()" and not "addView()", b.c. "getView()" is the public, generic way to get the view and "addView()" is the annotator-specific implementation, and I wanted a uniform mechanism.... I changed a number of Annotators (POS, Chunker, NER, Lemmatizer, SimpleGazetteerAnnotator) to use lazy init by default, and added some basic tests.

### CODE REVIEW

File: edison/src/test/java/edu/illinois/cs/cogcomp/edison/annotators/SimpleGazetteerAnnotatorTest.java [hunk 1]
```diff
@@ -211,26 +227,30 @@ public void testAddView() throws IOException, URISyntaxException, AnnotatorExcep
...
-	 * @throws URISyntaxException 
-	 * @throws IOException 
-	 */
-	@Test
-	public void testSimpleGazetteerAnnotatorIntString() throws IOException, URISyntaxException {
-		SimpleGazetteerAnnotator sga = new SimpleGazetteerAnnotator(6, "/testgazetteers/");
-		assertTrue ("Wrong number of dictionaries loaded.",sga.dictionaries.size() == 1);
-		assertTrue ("Wrong number of dictionaries loaded.",sga.dictionariesIgnoreCase.size() == 1);
-	}
+//	/**
```
Comment (danyaljj) [inline edison/src/test/java/edu/illinois/cs/cogcomp/edison/annotators/SimpleGazetteerAnnotatorTest.java]: Hopefully we don't want to keep this commented out?

Comment (mssammon) [inline edison/src/test/java/edu/illinois/cs/cogcomp/edison/annotators/SimpleGazetteerAnnotatorTest.java]: It should now be superfluous as the lazy init pattern as currently implemented requires a ResourceManager to be passed up to the Annotator, so different parameter configurations are accommodated by changing the RM entries.

**Comment by mssammon** [comment]
@cowchipkid please take a look at these changes and comment as needed; once you're happy, you can go ahead and merge unless Daniel sees other problems.  I can go ahead and delete the commented test if you agree.

**Comment by cowchipkid** [comment]
IllinoisLemmatizer appears to circumvent this pattern, it calls initialize explicitly in some constructors, but not others, and it seems to be the only class to set the initialized flag bypassing the superclasses mechanism to handle initialization. How come?

**Comment by mssammon** [comment]
@cowchipkid can you elaborate please?  The default constructor uses default flag 'true' for lazy initialization, but user can call a constructor with an explicit flag to set non-lazy, or use the constructor with ResourceManager to specify it.

**Comment by mssammon** [comment]
This would require two initialize() methods, I think: one in Annotator that calls an abstract sub-class initialize() method, then sets the "isInitialized" flag.  Acceptable?

**Comment by cowchipkid** [comment]
Sure, but I think I would name the method in the superclass doInitialize or some such.

**Comment by cowchipkid** [comment]
Semaphor seems unhappy...

**Comment by danyaljj** [comment]
``` 
testCreateWnLemmaView(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest)  Time elapsed: 0.008 sec  <<< ERROR!
java.lang.IllegalStateException: Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:183)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:154) 
```

**Comment by mssammon** [comment]
I'm baffled. This doesn't show up on my local machine.

**Comment by mssammon** [comment]
I think the earlier problems to do with branch management issues on my end.  Current failure is due to the large dependency issue, doesn't reach the previous point of failure so impossible to tell if problem is fixed. Someone please do me a favor: download the branch and install it and report back...

**Comment by mssammon** [comment]
found a probably unrelated bug for Annotators that try to log during non-lazy initialize() -- fixing

**Comment by mssammon** [comment]
This is now working in another project (pipeline) via a snapshot deployment...

**Comment by mssammon** [comment]
OK spoke too soon. Returning to WIP status.

**Comment by mssammon** [comment]
This should be ready to go now -- the problems I was having were to do with IntelliJ holding on to an outdated snapshot somehow...

**Comment by danyaljj** [comment]
I think lemmatizer tests are unhappy on semaphore: 
``` 
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building illinois-lemmatizer 3.0.58
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ illinois-lemmatizer ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] 
[INFO] --- maven-compiler-plugin:2.0.2:compile (default-compile) @ illinois-lemmatizer ---
[INFO] Compiling 12 source files to /home/runner/illinois-cogcomp-nlp/lemmatizer/target/classes
ngMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

Running edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerPlainTest
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.001 sec <<< FAILURE!
java.lang.IllegalStateException: Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:183)test(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerPlainTest)  Time elapsed: 0.001 sec  <<< ERROR!

	at edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerPlainTest.test(LemmatizerPlainTest.java:46)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	
		at org.junit.internal.runners.TestMethod.invoke(TestMethod.java:59)at org.junit.internal.runners.MethodRoadie.runTestMethod(MethodRoadie.java:98)
at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.internal.runners.MethodRoadie$2.run(MethodRoadie.java:79)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:77)Tests run: 4, Failures: 0, Errors: 3, Skipped: 0, Time elapsed: 0.598 sec <<< FAILURE!
testCreateWnLemmaView(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest)  Time elapsed: 0.059 sec  <<< ERROR!
java.lang.IllegalStateException: Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:183)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:154)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.createLemmaView(IllinoisLemmatizer.java:280)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest.testCreateWnLemmaView(LemmatizerTATest.java:81)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.internal.runners.TestMethod.invoke(TestMethod.java:59)
	at org.junit.internal.runners.MethodRoadie.runTestMethod(MethodRoadie.java:98)
	at org.junit.internal.runners.MethodRoadie$2.run(MethodRoadie.java:79)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:77)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:42)
	at org.junit.internal.runners.JUnit4ClassRunner.invokeTestMethod(JUnit4ClassRunner.java:88)
	at org.junit.internal.runners.JUnit4ClassRunner.runMethods(JUnit4ClassRunner.java:51)
	at org.junit.internal.runners.JUnit4ClassRunner$1.run(JUnit4ClassRunner.java:44)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:27)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:37)
	at org.junit.internal.runners.JUnit4ClassRunner.run(JUnit4ClassRunner.java:42)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

simpleTest(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest)  Time elapsed: 0.01 sec  <<< ERROR!
java.lang.IllegalStateException: Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:183)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest.simpleTest(LemmatizerTATest.java:40)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.internal.runners.TestMethod.invoke(TestMethod.java:59)
	at org.junit.internal.runners.MethodRoadie.runTestMethod(MethodRoadie.java:98)
	at org.junit.internal.runners.MethodRoadie$2.run(MethodRoadie.java:79)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:77)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:42)
	at org.junit.internal.runners.JUnit4ClassRunner.invokeTestMethod(JUnit4ClassRunner.java:88)
	at org.junit.internal.runners.JUnit4ClassRunner.runMethods(JUnit4ClassRunner.java:51)
	at org.junit.internal.runners.JUnit4ClassRunner$1.run(JUnit4ClassRunner.java:44)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:27)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:37)
	at org.junit.internal.runners.JUnit4ClassRunner.run(JUnit4ClassRunner.java:42)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

testCreateTextAnnotationLemmaView(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest)  Time elapsed: 0.008 sec  <<< ERROR!
java.lang.IllegalStateException: Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:183)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.getLemma(IllinoisLemmatizer.java:154)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.IllinoisLemmatizer.createLemmaView(IllinoisLemmatizer.java:280)
	at edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest.testCreateTextAnnotationLemmaView(LemmatizerTATest.java:126)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.internal.runners.TestMethod.invoke(TestMethod.java:59)
	at org.junit.internal.runners.MethodRoadie.runTestMethod(MethodRoadie.java:98)
	at org.junit.internal.runners.MethodRoadie$2.run(MethodRoadie.java:79)
	at org.junit.internal.runners.MethodRoadie.runBeforesThenTestThenAfters(MethodRoadie.java:87)
	at org.junit.internal.runners.MethodRoadie.runTest(MethodRoadie.java:77)
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:42)
	at org.junit.internal.runners.JUnit4ClassRunner.invokeTestMethod(JUnit4ClassRunner.java:88)
	at org.junit.internal.runners.JUnit4ClassRunner.runMethods(JUnit4ClassRunner.java:51)
	at org.junit.internal.runners.JUnit4ClassRunner$1.run(JUnit4ClassRunner.java:44)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:27)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:37)
	at org.junit.internal.runners.JUnit4ClassRunner.run(JUnit4ClassRunner.java:42)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.Delegati
	at org.junit.internal.runners.MethodRoadie.run(MethodRoadie.java:42)
	at org.junit.internal.runners.JUnit4ClassRunner.invokeTestMethod(JUnit4ClassRunner.java:88)
	at org.junit.internal.runners.JUnit4ClassRunner.runMethods(JUnit4ClassRunner.java:51)
	at org.junit.internal.runners.JUnit4ClassRunner$1.run(JUnit4ClassRunner.java:44)
	at org.junit.internal.runners.ClassRoadie.runUnprotected(ClassRoadie.java:27)
	at org.junit.internal.runners.ClassRoadie.runProtected(ClassRoadie.java:37)
	at org.junit.internal.runners.JUnit4ClassRunner.run(JUnit4ClassRunner.java:42)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

Tests in error: 


Results :
  testCreateWnLemmaView(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest): Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
  simpleTest(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest): Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
  test(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerPlainTest): Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.
  testCreateTextAnnotationLemmaView(edu.illinois.cs.cogcomp.nlp.lemmatizer.LemmatizerTATest): Not initialized! are you using lazy initialization? Need to call 'initialize()' before calling this method.

```

**Comment by mssammon** [comment]
OK: pushed from correct branch this time...

**Comment by mssammon** [comment]
Just found another unrelated bug in DummyTextAnnotator ('+' instead of ',' in lemma view source array meant one fewer token than other views...) fixing now and will push again after tests pass locally

**Comment by cowchipkid** [comment]
Do you want me to merge this? I don’t seem to have permissions to do that, or at least, I don’t see the merge button anywhere…

**Pull request merged**